ci: Enable DEBUG=1 for one GCC-12+ build to catch 117966 regressions

maflcko opened the pull request:
It is possible that someone accidentally removes the workaround in fa9e0489f57968945d54ef56b275f51540f3e5e4, or more likely that someone accidentally adds new code without the workaround.

Avoid this by adding a temporary CI check.

This can be tested by reverting the workaround and observing a failure.

Commit:
ci: Enable DEBUG=1 for one GCC-12+ build to catch 117966 regressions

hebasto reviewed (approved):
ACK fa1d84702bf4c14d55c51b9ab4cd30cdbbc5ad44, tested locally by reverting https://github.com/bitcoin/bitcoin/commit/fa9e0489f57968945d54ef56b275f51540f3e5e4 back and observing a compiler error.

fanquake commented:
One thought I have here is that not only are we no-longer running any unit tests for the Win release builds in the CI, but after this change, the build will also no-longer be a release build, so we are moving even further away from testing in the CI, what we are actually producing in Guix. I guess if this is temporary, maybe it's ok, but worth noting, and if it's possible to test this in a different way, without moving the Windows build away from being a release build, it might be worthwhile.

fanquake commented:
https://cirrus-ci.com/task/5812783272427520?logs=ci#L4621:
```bash
[15:22:49.280] [ 86%] Building CXX object src/test/fuzz/CMakeFiles/fuzz.dir/float.cpp.o
[15:22:49.281] cd /ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu/src/test/fuzz && /usr/bin/ccache /opt/rh/gcc-toolset-12/root/usr/bin/g++ -m64 -DABORT_ON_FAILED_ASSUME -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DDEBUG -DDEBUG_LOCKCONTENTION -DDEBUG_LOCKORDER -DRPC_DOC_CHECK -D_GLIBCXX_DEBUG -D_GLIBCXX_DEBUG_PEDANTIC -D_LIBCPP_HARDENING_MODE=_LIBCPP_HARDENING_MODE_DEBUG -I/ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu/src -I/ci_container_base/src -I/ci_container_base/src/univalue/include -I/ci_container_base/src/minisketch/include -I/ci_container_base/src/leveldb/include -I/ci_container_base/src/secp256k1/include -isystem /ci_container_base/depends/x86_64-pc-linux-gnu/include -pipe -std=c++20 -O0 -ftrapv -O1 -g3 -g3 -std=c++20 -fPIE -fvisibility=hidden -fno-extended-identifiers -fdebug-prefix-map=/ci_container_base/src=. -fmacro-prefix-map=/ci_container_base/src=. -fstack-reuse=none -Werror -Wall -Wextra -Wformat -Wformat-security -Wvla -Wredundant-decls -Wdate-time -Wduplicated-branches -Wduplicated-cond -Wlogical-op -Woverloaded-virtual -Wsuggest-override -Wimplicit-fallthrough -Wunreachable-code -Wbidi-chars=any -Wundef -Wno-unused-parameter -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -MD -MT src/test/fuzz/CMakeFiles/fuzz.dir/float.cpp.o -MF CMakeFiles/fuzz.dir/float.cpp.o.d -o CMakeFiles/fuzz.dir/float.cpp.o -c /ci_container_base/src/test/fuzz/float.cpp  
[15:22:51.900] In file included from /ci_container_base/src/test/fuzz/float.cpp:12:
[15:22:51.900] In function ‘constexpr bool std::isnan(double)’,
[15:22:51.900]     inlined from ‘void float_fuzz_target(FuzzBufferType)’ at /ci_container_base/src/test/fuzz/float.cpp:58:9:
[15:22:51.900] /opt/rh/gcc-toolset-12/root/usr/include/c++/12/cmath:620:27: error: ‘tmp’ may be used uninitialized [-Werror=maybe-uninitialized]
[15:22:51.900]   620 |   { return __builtin_isnan(__x); }
[15:22:51.900]       |            ~~~~~~~~~~~~~~~^~~~~
[15:22:51.900] /ci_container_base/src/test/fuzz/float.cpp: In function ‘void float_fuzz_target(FuzzBufferType)’:
[15:22:51.900] /ci_container_base/src/test/fuzz/float.cpp:21:20: note: ‘tmp’ was declared here
[15:22:51.900]    21 |             double tmp;
[15:22:51.900]       |                    ^~~
[15:22:51.900] cc1plus: all warnings being treated as errors
```

Guess it's possible this is solved after #31593, so we could do that first.

Commit:
refactor: Avoid GCC false positive error

This avoids an overly agressive GCC false positive warning:

error: ‘tmp’ may be used uninitialized [-Werror=maybe-uninitialized]

hebasto reviewed (approved):
ACK fa8ade300f421dcc3b0cd956ab03a50a5ae80646, I've tested locally on Ubuntu 24.04.

fa8ade300f421dcc3b0cd956ab03a50a5ae80646 -- I faced the same false-positive on Alpine.
